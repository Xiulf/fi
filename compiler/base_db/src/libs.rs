use std::fmt;
use std::ops::Index;
use std::path::PathBuf;

use cfg::CfgOptions;
use rustc_hash::{FxHashMap, FxHashSet};
use vfs::FileId;

#[derive(Debug, Clone)]
pub struct LibData {
    pub id: LibId,
    pub name: String,
    pub kind: LibKind,
    pub deps: Vec<LibId>,
    pub links: Vec<PathBuf>,
    pub cfg_options: CfgOptions,

    /// Shade projects don't have a root file
    /// However, source roots are only created after loading projects
    /// so this is used to identify the project's source root.
    pub root_file: FileId,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LibKind {
    Dynamic,
    Static,
    Executable,
}

impl Default for LibKind {
    fn default() -> Self {
        LibKind::Dynamic
    }
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct LibId(pub u32);

#[derive(Default, Debug, Clone)]
pub struct LibSet {
    libs: FxHashMap<LibId, LibData>,
}

impl LibSet {
    pub fn add_lib(
        &mut self,
        name: impl Into<String>,
        kind: LibKind,
        links: Vec<PathBuf>,
        cfg_options: CfgOptions,
        root_file: FileId,
    ) -> LibId {
        let name = name.into();
        let id = LibId(self.libs.len() as u32);
        let data = LibData {
            id,
            name,
            kind,
            links,
            cfg_options,
            root_file,
            deps: Vec::new(),
        };

        self.libs.insert(id, data);
        id
    }

    pub fn add_dep(&mut self, from: LibId, to: LibId, cfg: &CfgOptions) -> Result<(), CyclicDependenciesError> {
        if self.dfs_find(from, to, &mut FxHashSet::default()) {
            return Err(CyclicDependenciesError {
                from: (from, self[from].name.clone()),
                to: (to, self[to].name.clone()),
            });
        }

        self.libs.get_mut(&from).unwrap().deps.push(to);

        let to = self.libs.get_mut(&to).unwrap();

        to.cfg_options = to.cfg_options.merge(cfg);

        Ok(())
    }

    pub fn extend(&mut self, other: LibSet) {
        self.libs.extend(other.libs);
    }

    pub fn is_empty(&self) -> bool {
        self.libs.is_empty()
    }

    pub fn iter(&self) -> impl Iterator<Item = LibId> + '_ {
        self.libs.keys().copied()
    }

    pub fn find(&self, name: &str) -> Option<&LibData> {
        self.libs.values().find(|l| l.name == name)
    }

    pub fn all_deps(&self, lib: LibId) -> Vec<LibId> {
        let mut res = Vec::new();
        let mut visited = FxHashSet::default();

        go(self, &mut visited, &mut res, lib);

        return res;

        fn go(libs: &LibSet, visited: &mut FxHashSet<LibId>, res: &mut Vec<LibId>, source: LibId) {
            if !visited.insert(source) {
                return;
            }

            for &dep in libs[source].deps.iter() {
                go(libs, visited, res, dep);
            }

            res.push(source);
        }
    }

    pub fn toposort(&self) -> Vec<LibId> {
        let mut res = Vec::new();
        let mut visited = FxHashSet::default();

        for &lib in self.libs.keys() {
            go(self, &mut visited, &mut res, lib);
        }

        return res;

        fn go(libs: &LibSet, visited: &mut FxHashSet<LibId>, res: &mut Vec<LibId>, source: LibId) {
            if !visited.insert(source) {
                return;
            }

            for &dep in libs[source].deps.iter() {
                go(libs, visited, res, dep);
            }

            res.push(source);
        }
    }

    fn dfs_find(&self, target: LibId, from: LibId, visited: &mut FxHashSet<LibId>) -> bool {
        if !visited.insert(from) {
            return false;
        }

        if target == from {
            return true;
        }

        for dep in &self[from].deps {
            if self.dfs_find(target, *dep, visited) {
                return true;
            }
        }

        false
    }
}

impl Index<LibId> for LibSet {
    type Output = LibData;

    fn index(&self, index: LibId) -> &Self::Output {
        &self.libs[&index]
    }
}

#[derive(Debug)]
pub struct CyclicDependenciesError {
    from: (LibId, String),
    to: (LibId, String),
}

impl fmt::Display for CyclicDependenciesError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "cyclid dependencies: {} -> {}", self.from.1, self.to.1)
    }
}

impl std::error::Error for CyclicDependenciesError {
}
