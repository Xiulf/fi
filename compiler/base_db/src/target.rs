use std::str::FromStr;

use target_lexicon::Triple;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompilerTarget {
    Javascript,
    Native(Triple),
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
pub enum Optimization {
    #[default]
    None,
}

impl Default for CompilerTarget {
    fn default() -> Self {
        Self::Native(Triple::host())
    }
}

impl CompilerTarget {
    const UNKNOWN: Triple = Triple {
        architecture: target_lexicon::Architecture::Unknown,
        vendor: target_lexicon::Vendor::Unknown,
        operating_system: target_lexicon::OperatingSystem::Unknown,
        environment: target_lexicon::Environment::Unknown,
        binary_format: target_lexicon::BinaryFormat::Unknown,
    };

    pub fn triple(&self) -> &Triple {
        match self {
            | Self::Native(triple) => triple,
            | _ => &Self::UNKNOWN,
        }
    }

    pub fn is_windows(&self) -> bool {
        match self {
            | Self::Native(triple) => triple.operating_system == target_lexicon::OperatingSystem::Windows,
            | _ => false,
        }
    }
}

impl FromStr for CompilerTarget {
    type Err = target_lexicon::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            | "js" | "javascript" => Ok(Self::Javascript),
            | _ => Triple::from_str(s).map(Self::Native),
        }
    }
}
