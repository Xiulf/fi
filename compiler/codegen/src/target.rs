use target_lexicon::{OperatingSystem, Triple};

mod linux_base;
mod windows_msvc_base;

#[derive(Debug, PartialEq, Eq)]
pub struct Target {
    pub triple: Triple,
    pub cpu: &'static str,
    pub features: &'static str,
    pub dll_prefix: &'static str,
    pub linker_flavor: LinkerFlavor,
    pub pre_link_args: Vec<&'static str>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LinkerFlavor {
    Ld,
    Msvc,
}

impl Default for Target {
    fn default() -> Self {
        Self {
            triple: Triple::unknown(),
            cpu: "generic",
            features: "",
            dll_prefix: "lib",
            linker_flavor: LinkerFlavor::Ld,
            pre_link_args: Vec::new(),
        }
    }
}

macro_rules! targets {
    ($(($triple:literal, $target:ident)),* $(,)?) => {
        $(mod $target;)*

        const TARGETS: &[&str] = &[$($triple),*];

        fn load_target(triple: &Triple) -> Target {
            match triple.to_string().as_str() {
                $($triple => return $target::target(),)*
                _ => {},
            }

            let linker_flavor = match triple.environment {
                target_lexicon::Environment::Msvc => LinkerFlavor::Msvc,
                _ => LinkerFlavor::Ld,
            };

            let dll_prefix = match triple.operating_system {
                target_lexicon::OperatingSystem::Windows => "",
                _ => "lib",
            };

            Target {
                triple: triple.clone(),
                dll_prefix,
                linker_flavor,
                ..Default::default()
            }
        }

        pub fn get_targets() -> &'static [&'static str] {
            TARGETS
        }
    };
}

targets! {
    ("x86_64-pc-windows-msvc", x86_64_pc_windows_msvc),
    ("x86_64-unknown-linux-gnu", x86_64_unknown_linux_gnu),
}

impl Target {
    pub fn load(triple: &Triple) -> Target {
        load_target(triple)
    }

    pub fn host_target() -> Target {
        Self::load(&Triple::host())
    }

    pub fn is_windows(&self) -> bool {
        matches!(self.triple.operating_system, OperatingSystem::Windows)
    }
}
