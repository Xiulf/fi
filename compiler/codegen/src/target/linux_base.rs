use target_lexicon::{Architecture, BinaryFormat, Environment, OperatingSystem, Triple, Vendor};

use super::{LinkerFlavor, Target};

pub fn opts() -> Target {
    Target {
        triple: Triple {
            architecture: Architecture::Unknown,
            vendor: Vendor::Unknown,
            operating_system: OperatingSystem::Linux,
            environment: Environment::Gnu,
            binary_format: BinaryFormat::Elf,
        },
        linker_flavor: LinkerFlavor::Ld,
        ..Default::default()
    }
}
