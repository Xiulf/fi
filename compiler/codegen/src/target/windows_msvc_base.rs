use target_lexicon::{Architecture, BinaryFormat, Environment, OperatingSystem, Triple, Vendor};

use super::{LinkerFlavor, Target};

pub fn opts() -> Target {
    Target {
        triple: Triple {
            architecture: Architecture::Unknown,
            vendor: Vendor::Pc,
            operating_system: OperatingSystem::Windows,
            environment: Environment::Msvc,
            binary_format: BinaryFormat::Coff,
        },
        linker_flavor: LinkerFlavor::Msvc,
        dll_prefix: "",
        ..Default::default()
    }
}
