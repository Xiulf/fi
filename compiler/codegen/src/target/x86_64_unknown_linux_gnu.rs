use target_lexicon::Architecture;

use super::Target;

pub fn target() -> Target {
    let mut target = Target {
        cpu: "x86-64",
        ..super::linux_base::opts()
    };

    target.triple.architecture = Architecture::X86_64;
    target
}
