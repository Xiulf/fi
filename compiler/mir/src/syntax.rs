use arena::{Arena, Idx};
use hir::id::DefWithBodyId;
use hir::{Ctor, DefWithBody, ExprId, Func};
use rustc_hash::FxHashMap;
use salsa::{InternId, InternKey};

use crate::repr::Repr;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Module {
    pub hir: hir::Module,
    pub functions: FxHashMap<Func, Function>,
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Linkage {
    #[default]
    Local,
    Import,
    Export,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct Function {
    pub linkage: Linkage,
    pub name: String,
    pub body: Option<Body>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Body(InternId);

impl InternKey for Body {
    fn from_intern_id(v: InternId) -> Self {
        Self(v)
    }

    fn as_intern_id(&self) -> InternId {
        self.0
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BodyOrigin {
    pub def: DefWithBodyId,
    pub expr: Option<ExprId>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct BodyData {
    pub origin: BodyOrigin,
    pub locals: Arena<LocalData>,
    pub blocks: Arena<BlockData>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Local(pub Idx<LocalData>);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct LocalData {
    pub id: Local,
    pub kind: LocalKind,
    pub repr: Repr,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LocalKind {
    Arg,
    Var,
    Tmp,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Block(pub Idx<BlockData>);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct BlockData {
    pub id: Block,
    pub params: Vec<Local>,
    pub stmts: Vec<Stmt>,
    pub term: Term,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Term {
    /// No terminator, must be replaced before codegen.
    None,

    /// Indicates that this code path should never be reached.
    Unreachable,

    /// Stop execution and exit the current process.
    Abort,

    /// Return the given value.
    Return(Operand),

    /// Jump to the given target.
    Jump(JumpTarget),

    /// switch over the given discriminant value.
    Switch {
        discr: Operand,
        values: Vec<i128>,
        targets: Vec<JumpTarget>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct JumpTarget {
    pub block: Block,
    pub args: Vec<Operand>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Stmt {
    /// Creates a new instance of the type for the given local.
    /// Does nothing for trivial types.
    Init(Local),

    /// Deletes the value of the given local.
    /// Does nothing for trivial types.
    Drop(Local),

    /// Assign a value to the given place.
    Assign(Place, Rvalue),

    /// Set the discriminant value of the place.
    SetDiscriminant(Place, Ctor),

    /// Call the given function with the arguments and store the result in place.
    Call {
        place: Place,
        func: Operand,
        args: Vec<Operand>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Rvalue {
    /// Use the value of the given operand.
    Use(Operand),

    /// Get the address of the place.
    Ref(Place),

    /// Get the discriminant value of the place.
    Discriminant(Place),

    /// Cast to a different type.
    Cast(Operand),

    /// Create a reference to the given body.
    BodyRef(Body),

    /// Create a reference to the given def.
    DefRef(DefWithBody),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Operand {
    /// Load the value at the place.
    /// This may perform extra operations for non-trivial types.
    Copy(Place),

    /// Load the value at the place.
    Move(Place),

    /// A constant value.
    Const(Const, Repr),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Place {
    pub local: Local,
    pub projection: Vec<Projection>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Projection {
    /// Dereference the pointer.
    Deref,

    /// Get a reference to a specific field of this place.
    Field(usize),

    /// Downcast the type of this place to the given variant.
    Downcast(Ctor),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Const {
    Undefined,
    Unit,
    Int(i128),
    Float(u64),
    Char(char),
    String(String),
    Ctor(Ctor),
}

impl Place {
    pub fn has_deref(&self) -> bool {
        self.projection
            .first()
            .map(|f| matches!(f, Projection::Deref))
            .unwrap_or(false)
    }
}
