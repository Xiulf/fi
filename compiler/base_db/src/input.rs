use relative_path::{RelativePath, RelativePathBuf};
use rustc_hash::{FxHashMap, FxHashSet};
use syntax::TextSize;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct FileId(pub u32);

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SourceRootId(pub u32);

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceRoot {
    pub is_lib: bool,
    files: FxHashMap<FileId, RelativePathBuf>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FileTree {
    root: FileId,
    children: FxHashMap<FileId, Vec<FileId>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LineIndex {
    newlines: Vec<TextSize>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct LineCol {
    pub line: u32,
    pub col: u32,
}

impl SourceRoot {
    pub fn new_local() -> Self {
        SourceRoot {
            is_lib: false,
            files: FxHashMap::default(),
        }
    }

    pub fn new_library() -> Self {
        SourceRoot {
            is_lib: true,
            files: FxHashMap::default(),
        }
    }

    pub fn insert_file(&mut self, file_id: FileId, path: impl AsRef<RelativePath>) {
        self.files.insert(file_id, path.as_ref().to_relative_path_buf());
    }

    pub fn remove_file(&mut self, file_id: FileId) -> bool {
        self.files.remove(&file_id).is_some()
    }

    pub fn relative_path(&self, file_id: FileId) -> &RelativePath {
        &self.files[&file_id]
    }

    pub fn files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.files.keys().copied()
    }

    pub fn iter(&self) -> impl Iterator<Item = (FileId, &RelativePath)> + '_ {
        self.files.iter().map(|(k, v)| (*k, v.as_relative_path()))
    }
}

impl FileTree {
    pub fn file_tree_query(db: &dyn crate::SourceDatabaseExt, lib: crate::libs::LibId) -> std::sync::Arc<Self> {
        let data = &db.libs()[lib];
        let source_root = db.source_root(data.source_root);
        let root = data.root_file;
        let mut tree = FileTree {
            root,
            children: FxHashMap::default(),
        };

        go(&mut tree, root, true, &source_root, &mut FxHashSet::default());

        return tree.into();

        fn go(tree: &mut FileTree, file: FileId, is_root: bool, root: &SourceRoot, visited: &mut FxHashSet<FileId>) {
            let path = root.relative_path(file);
            let mut children = Vec::new();

            visited.insert(file);

            for (file_id, file_path) in root.iter() {
                if visited.contains(&file_id) {
                    continue;
                }

                if file_path.parent() == Some(RelativePath::new("")) && is_root {
                    children.push(file_id);
                    go(tree, file_id, false, root, visited);
                } else if file_path.parent() == Some(&*path.with_file_name(path.file_stem().unwrap())) {
                    children.push(file_id);
                    go(tree, file_id, false, root, visited);
                }
            }

            tree.children.insert(file, children);
        }
    }

    pub fn root(&self) -> FileId {
        self.root
    }

    pub fn children(&self, file: FileId) -> impl Iterator<Item = FileId> + '_ {
        self.children[&file].iter().copied()
    }
}

impl LineIndex {
    pub fn new(text: &str) -> Self {
        let mut newlines = vec![TextSize::from(0)];
        let mut curr_row = TextSize::from(0);

        for c in text.chars() {
            curr_row += TextSize::of(c);

            if c == '\n' {
                newlines.push(curr_row);
            }
        }

        LineIndex { newlines }
    }

    pub fn line_col(&self, offset: TextSize) -> LineCol {
        let line = self
            .newlines
            .iter()
            .position(|l| l > &offset)
            .unwrap_or(self.newlines.len())
            - 1;

        let line_start_offset = self.newlines[line];
        let col = offset - line_start_offset;

        LineCol {
            line: line as u32,
            col: col.into(),
        }
    }

    pub fn offset(&self, line_col: LineCol) -> TextSize {
        // @TODO: modify this to account for characters larger than 1 byte
        self.newlines[line_col.line as usize] + TextSize::from(line_col.col)
    }

    pub fn line_offset(&self, line: u32) -> usize {
        self.newlines[line as usize].into()
    }
}
