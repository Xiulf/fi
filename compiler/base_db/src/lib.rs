pub mod input;
pub mod libs;
pub mod target;

use std::sync::Arc;
use std::{fmt, panic};

use input::{FileId, LineIndex, SourceRoot, SourceRootId};
use libs::LibId;
use rustc_hash::FxHashSet;
pub use salsa::Cancelled;
use syntax::{ast, Parsed};
use target::CompilerTarget;

pub trait Upcast<T: ?Sized> {
    fn upcast(&self) -> &T;
}

#[salsa::query_group(SourceDatabaseStorage)]
pub trait SourceDatabase: FileLoader {
    fn parse(&self, file_id: FileId) -> Parsed<ast::SourceFile>;

    #[salsa::input]
    fn target(&self) -> CompilerTarget;

    #[salsa::input]
    fn libs(&self) -> Arc<libs::LibSet>;

    #[salsa::input]
    fn file_source_root(&self, file_id: FileId) -> SourceRootId;

    #[salsa::input]
    fn source_root(&self, id: SourceRootId) -> Arc<SourceRoot>;
}

#[salsa::query_group(SourceDatabaseExtStorage)]
pub trait SourceDatabaseExt: SourceDatabase {
    #[salsa::input]
    fn file_text(&self, file_id: FileId) -> Arc<str>;

    fn source_root_libs(&self, id: SourceRootId) -> Arc<FxHashSet<LibId>>;

    fn line_index(&self, file_id: FileId) -> Arc<LineIndex>;
}

fn parse(db: &dyn SourceDatabase, file_id: FileId) -> Parsed<ast::SourceFile> {
    let text = db.file_text(file_id);

    ast::SourceFile::parse(&*text)
}

fn source_root_libs(db: &dyn SourceDatabaseExt, id: SourceRootId) -> Arc<FxHashSet<LibId>> {
    let libs = db.libs();

    Arc::new(
        libs.iter()
            .filter(|&lib| db.file_source_root(libs[lib].root_file) == id)
            .collect(),
    )
}

fn line_index(db: &dyn SourceDatabaseExt, file_id: FileId) -> Arc<LineIndex> {
    let text = SourceDatabaseExt::file_text(db, file_id);

    Arc::new(LineIndex::new(&text))
}

pub trait FileLoader {
    fn file_text(&self, file_id: FileId) -> Arc<str>;
    fn relevant_libs(&self, file_id: FileId) -> Arc<FxHashSet<LibId>>;
}

pub struct FileLoaderDelegate<T>(pub T);

impl<T: SourceDatabaseExt> FileLoader for FileLoaderDelegate<&'_ T> {
    fn file_text(&self, file_id: FileId) -> Arc<str> {
        SourceDatabaseExt::file_text(self.0, file_id)
    }

    fn relevant_libs(&self, file_id: FileId) -> Arc<FxHashSet<LibId>> {
        let source_root = self.0.file_source_root(file_id);

        self.0.source_root_libs(source_root)
    }
}

#[derive(Debug)]
pub struct ICE(pub std::borrow::Cow<'static, str>);

impl ICE {
    pub fn throw(msg: impl Into<std::borrow::Cow<'static, str>>) -> ! {
        panic::panic_any(Self(msg.into()))
    }
}

impl fmt::Display for ICE {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "internal compiler error: '{}'", self.0)
    }
}

impl std::error::Error for ICE {
}

#[derive(Debug)]
pub struct Error(pub std::borrow::Cow<'static, str>);

impl Error {
    pub fn throw(msg: impl Into<std::borrow::Cow<'static, str>>) -> ! {
        panic::panic_any(Self(msg.into()))
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "error: '{}'", self.0)
    }
}

impl std::error::Error for Error {
}
