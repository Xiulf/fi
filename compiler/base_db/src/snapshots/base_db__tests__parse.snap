---
source: compiler/base_db/src/lib.rs
expression: node
---
([
  Event(
    t: "EnterNode",
    c: (57, false),
  ),
  Event(
    t: "EnterNode",
    c: (64, false),
  ),
  Event(
    t: "Token",
    c: (30, "module"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "EnterNode",
    c: (84, false),
  ),
  Event(
    t: "EnterNode",
    c: (85, false),
  ),
  Event(
    t: "EnterNode",
    c: (87, false),
  ),
  Event(
    t: "Token",
    c: (7, "Core"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (19, "."),
  ),
  Event(
    t: "EnterNode",
    c: (85, false),
  ),
  Event(
    t: "EnterNode",
    c: (87, false),
  ),
  Event(
    t: "Token",
    c: (7, "Cmp"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (4, "\n\n"),
  ),
  Event(
    t: "EnterNode",
    c: (67, false),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (6, "main"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (9, "0"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (4, "\n\n"),
  ),
  Event(
    t: "EnterNode",
    c: (68, false),
  ),
  Event(
    t: "Token",
    c: (41, "type"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (7, "X"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (3, "\n    "),
  ),
  Event(
    t: "EnterNode",
    c: (77, false),
  ),
  Event(
    t: "Token",
    c: (27, "|"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (7, "Y"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (4, "\n    "),
  ),
  Event(
    t: "EnterNode",
    c: (77, false),
  ),
  Event(
    t: "Token",
    c: (27, "|"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (7, "Z"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (5, "\n\n"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (4, ""),
  ),
  Event(
    t: "EnterNode",
    c: (69, false),
  ),
  Event(
    t: "Token",
    c: (42, "trait"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (7, "Iterator"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (6, "self"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (6, "it"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (3, "\n    "),
  ),
  Event(
    t: "EnterNode",
    c: (67, false),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (6, "next"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (25, "::"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (6, "self"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (23, "->"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (7, "Option"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (6, "it"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (5, "\n\n"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (4, ""),
  ),
  Event(
    t: "EnterNode",
    c: (70, false),
  ),
  Event(
    t: "Token",
    c: (43, "impl"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (7, "Iterator"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (7, "Iter"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (7, "Item"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (3, "\n    "),
  ),
  Event(
    t: "EnterNode",
    c: (67, false),
  ),
  Event(
    t: "EnterNode",
    c: (86, false),
  ),
  Event(
    t: "Token",
    c: (6, "next"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (6, "self"),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (21, "="),
  ),
  Event(
    t: "Token",
    c: (1, " "),
  ),
  Event(
    t: "Token",
    c: (22, "_"),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (5, ""),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "Token",
    c: (4, ""),
  ),
  Event(
    t: "LeaveNode",
  ),
  Event(
    t: "LeaveNode",
  ),
], [])
